slf4j/src/main/java/com/google/common/flogger/backend/slf4j/Slf4jLoggerBackend.java

```diff
@@ -0,0 +1,201 @@
+import com.google.common.flogger.backend.SimpleMessageFormatter;
+import org.slf4j.Logger;
+
+import java.util.logging.Level;
+
+/**
+ * A logging backend that uses slf4j to output log statements.
+ */
+final class Slf4jLoggerBackend extends LoggerBackend implements
+    SimpleMessageFormatter.SimpleLogHandler {
+
+  private final Logger logger;
+
+  private static final MappedLevel OFF_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return false;
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      // no-op
+    }
+  });
+
+  private static final MappedLevel TRACE_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isTraceEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.trace(message, thrown);
+    }
+  });
+
+  private static final MappedLevel DEBUG_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isDebugEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.debug(message, thrown);
+    }
+  });
+  private static final MappedLevel INFO_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isInfoEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.info(message, thrown);
+    }
+  });
+  private static final MappedLevel WARN_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isWarnEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.warn(message, thrown);
+    }
+  });
+  private static final MappedLevel ERROR_LEVEL = new MappedLevel(new MappedLevel.Predicate() {
+
+    @Override
+    public boolean test(Logger logger) {
+      return logger.isErrorEnabled();
+    }
+  }, new MappedLevel.LogAdapter() {
+    @Override
+    public void log(Logger logger, String message, Throwable thrown) {
+      logger.error(message, thrown);
+    }
+  });
+
+  Slf4jLoggerBackend(Logger logger) {
+    if (logger == null) {
+      throw new NullPointerException("logger is required");
+    }
+    this.logger = logger;
+  }
+
+  private static MappedLevel mapLevel(Level level) {
+    int requestedLevel = level.intValue();
+    if( requestedLevel < Level.FINE.intValue() ) {
+      return TRACE_LEVEL;
+    }
+
+    if( requestedLevel < Level.INFO.intValue() ) {
```

Comment:
Would you mind tidying up formatting while I'm giving this a look. I'm still not sure it needs to be as structurally complex as it does (e.g. why is MappedLevel a class at all?), since there are so few comments. If you added JavaDoc and/or code comments explaining why you've taken the approach you did, it'd help me a lot in the review.
Thanks.